### ISSUE TITLE
soup adapters audit does not check the learning rate, though MLX already records peak_lr

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#942]
## What

`soup adapters audit` (#762 / #763) compares what a finished run recorded against the config that started it. It does not audit the learning rate.

`training.lr` is arguably the most consequential setting in a training config — it is the one that decides whether a run was worth its GPU hours — and it is exactly the kind of value that can differ between what the user wrote and what the trainer used, because several code paths derive from it rather than passing it through.

On the MLX backend the derived value is already recorded and is right there to compare against:

```
src/soup_cli/trainer/mlx_optim.py:77    peak_lr: float
src/soup_cli/trainer/mlx_optim.py:88        "peak_lr": self.peak_lr,
src/soup_cli/trainer/mlx_optim.py:165       peak_lr=float(lr),
src/soup_cli/trainer/mlx_optim.py:177       peak, warmup, total = plan.peak_lr, ...
```

So the audit has a recorded number and a configured number and does not compare them.

## Why it is filed rather than left in a review thread

I called this "a natural follow-up" in review of #763 on 2026-09-12 and did not file it. A follow-up that lives only in a PR comment is one nobody can find, claim or close — which is the whole argument for the issue tracker, and I did not apply it to my own suggestion. Filing it now with the evidence attached.

This is **not** a change request on #763. That PR's own scope is fine and its two outstanding items are unrelated.

## Acceptance criteria

- [ ] The audit reports a mismatch between `training.lr` and the learning rate the run actually used, for at least the MLX path where `peak_lr` is already recorded.
- [ ] A test that fails when the two diverge — construct a recorded run whose `peak_lr` differs from its config's `lr` and assert the audit flags it. The mutation that must be caught: make the comparison always report equal.
- [ ] A decision, stated in the PR, about the **transformers** path: either audit the effective LR there too, or say plainly in the audit output that the LR was not checked on that backend. Silently auditing one backend and not the other is worse than auditing neither, because the clean report reads as coverage.
- [ ] Warmup and schedule interact with this. If the recorded `peak_lr` is by definition the post-warmup peak while `training.lr` is the configured base, say so — a "mismatch" that is actually correct behaviour is a false positive, and a guard that fires on correct runs gets deleted. Decide which number is the right comparand and write down why.

## Notes for whoever takes this

Start by measuring rather than coding: take one real MLX run, print the recorded `peak_lr` next to the config's `training.lr`, and see whether they are equal today. If they always are, this is a cheap guard against a future divergence; if they already differ for a legitimate reason, that reason is the actual content of this issue and the fix is different from what the title suggests.

Related: #762, #763 (the audit itself).

**Comment by kok-o** [comment]
@MakazhanAlpamys I'd like to work on this. I will add `training.lr` to `soup adapters audit` so it compares what the run recorded against the config that started it (including MLX `peak_lr`), and add regression tests for both matching and mismatched learning rate cases.

**Comment by Amix29** [comment]
I will take this. I will first measure a real local MLX training run on Apple Silicon and compare the recorded peak_lr with training.lr, then extend soup adapters audit so the result explicitly distinguishes checked MLX learning rates from backends where the effective LR was not recorded. I will add divergence and clean-path tests, including a mutation control that forces equality, and keep all heavy MLX imports lazy. No CUDA path is involved.

**Comment by Amix29** [comment]
Implementation and local validation are complete on `Amix29:codex/fix-942-audit-learning-rate` (delta commit `979b9bc`).

Measured first on a real native MLX/Metal smoke run as requested: `training.lr: 5e-4` produced `peak_lr: 0.0005` in the finished `adapter_config.json`. The audit now compares `training.lr` with the optimizer plan's `peak_lr`, deliberately not with the config echo `learning_rate`; warmup/cosine semantics are documented. PEFT/transformers reports the effective learning rate as explicitly not checked. Divergent, missing, malformed/non-finite, warmup, transformer, and mutation-control cases are covered; forcing equality makes the divergent test fail. Targeted result: 148 passed; full Ruff clean.

I am intentionally not opening a duplicated stacked PR while #763 is still open: the command and guard this change extends are not on `main` yet. The branch is based on #763 head `3f3e08a`; as soon as #763 merges I will replay only `979b9bc` onto current `main`, add the PR-numbered changelog fragment, and open the focused PR.

**Comment by MakazhanAlpamys** [comment]
Fixed in `87bba902` by @Amix29 (#1109): `soup adapters audit` now checks the effective MLX learning rate against the configured peak. Thank you.
